### ISSUE TITLE
[Feature request] MPEG-DASH Heartbeat

### ISSUE THREAD

**Issue by Frasknas** [peak3d/inputstream.adaptive#1736]
### Describe the feature

Hello.

I am writing a plugin for kodi to play streams from a play site that uses https://www.maginepro.com/ as their backend.

The streams are mpeg-dash. I have not worked with those before and I can't get inputstream.adaptive to play them correctly.
The stream will start to play but after about a minute it crashes.

The .json I retrieve from the website that contains information about a stream that I want to play looks like this:
`{
"playlist":"https://d1iohm9id1tl8o.cloudfront.net/out/v1/d2fc11689c1246728e722e26f21c2de7/index.mpd?start=1732465800&end=1732480200",
"license":"https://client-api.magine.com/api/playback/v1/widevine/license",
"heartbeat":"https://client-api.magine.com/api/playback/v1/heartbeat",
"headers":{
  "Magine-Play-Protocol":"dashs",
  "Magine-Play-DevicePlatform":"firefox",
  "Magine-Play-DeviceId":"8b08092f-1c24-4f88-ba06-....c1db1dec",
  "Magine-Play-Session":"fb710e1d-dbb3-4608-81e8-....efd14fba",
  "Magine-Play-DeviceModel":"firefox 104.0 / windows 10",
  "Magine-Play-DRM":"widevine",
  "Magine-Play-DeviceType":"web",
  "Magine-Play-EntitlementId":"eyJ0eXA.................ZgvEA"
  },
"playlistDuration":7200,
"reloadManifest":1732478309,
"rightsDenied":["offline"]
}`

I notice that there is a heartbeat URL. When I play the stream using the web interface and check the network traffic I can see that the browser sends a request to this URL every 10 seconds. My theory is that the reason my stream crash after a while is that I don't send any heartbeats.

I have looked at the documentation for inputstream.adaptive but I can't find any references to heartbeats.

Are heartbeats a standard feature in mpeg-dash or is it something that this unique for this site? Are they supported in your library or are there any plans to add support for them in a coming release?

I am attaching a log from one of my sessions that crashed.

Thanks in advance.
[kodi_crash_after_awhile.log](https://github.com/user-attachments/files/17940060/kodi_crash_after_awhile.log)

### Motivation

I can't get the streams to work without it

### Are you willing to contribute in development?

Yes

### Should a possible backport be considered?

No

**Comment by CastagnaIT** [comment]
I have already seen some services that implement some kind of "heartbeat", these are customizations, that regular dash dont need it, usually services use these callbacks to keep track of the current playback state of the user

if you want add the support to these callbacks must be done in your add-on and not on ISA, since dont exists a standard

i dont think that there is a direct relation about the missing "heartbeat" and the crash
unless your service deliberately blocks some data in the manifest, in which case you have to try check saved manifests

what i can see in the log is that after some time there are ffmpeg errors
this suggest that for some unknown reason the audio track become encrypted in a way that is not supported, or there is some kind of unsupported thing that cause bad decrypting, idk because in order to try understand something i would have to see all the saved manifest until the problem occurs, but for problems like this it might not be enough to determine what is happening

**Comment by CastagnaIT** [comment]
closed for no activity

**Comment by Frasknas** [comment]
Hello again.

I have been digging more into this. I now agree with you that the problem with the stream dying after a couple of minutes is probably not related to the heartbeats.

What I did notice is that the web app I am trying to emulate is sending multiple license requests. First one is sent when the stream is started. Then a new one is sent somewhere between a few seconds and up to two minutes after the initial request. After that it sends a new license request every 2 minutes as long as the stream is playing.

Since the life time of my stream is in the same ball park as the time until the web app sends the second license request my theory is that it dies because I am not sending these additional requests. 

I had a look at the source code. I found that CSession::OnSegmentChanged is a suitable place to inject code to test my theory. But I failed to find a way to send a new license POST without also recreating the stream. Is it possible?

Thanks in advance.
